frontend/src/lib/k8s/api/v2/webSocket.ts

```diff
@@ -1,11 +1,431 @@
+   * @param key - The unique subscription key
+   * @param onMessage - The message handler to remove
+   * @param userId - The user ID associated with the subscription
+   * @param path - The API path being watched
+   * @param query - Query parameters for the subscription
+   */
+  handleUnsubscribe(
+    key: string,
+    onMessage: (data: any) => void,
+    userId: string | null,
+    path: string,
+    query: string
+  ): void {
+    const listeners = this.listeners.get(key);
+    listeners?.delete(onMessage);
+
+    // Clean up when last listener is removed
+    if (listeners?.size === 0) {
+      this.listeners.delete(key);
+      this.completedPaths.delete(key);
+      this.activeSubscriptions.delete(key);
+
+      // Send close message if socket is open
+      if (this.socketMultiplexer?.readyState === WebSocket.OPEN) {
+        const [clusterId] = key.split(':');
+        const closeMsg: WebSocketMessage = {
+          clusterId,
+          path,
+          query,
+          userId: userId || '',
+          type: 'CLOSE',
+        };
+        this.socketMultiplexer.send(JSON.stringify(closeMsg));
+      }
+    }
+
+    // Close socket if no more listeners
+    if (this.listeners.size === 0) {
+      this.socketMultiplexer?.close();
+      this.socketMultiplexer = null;
+    }
+  },
+};
+
+/**
+ * React hook for WebSocket subscription to Kubernetes resources
+ * @param options - Configuration options for the WebSocket connection
+ * @param options.url - Function that returns the WebSocket URL to connect to
+ * @param options.enabled - Whether the WebSocket connection should be active
+ * @param options.cluster - The Kubernetes cluster ID to watch
+ * @param options.onMessage - Callback function to handle incoming Kubernetes events
+ * @param options.onError - Callback function to handle connection errors
+ *
+ * @example
+ * useWebSocket<Pod>({
+ *   url: () => '/api/v1/pods?watch=1',
+ *   enabled: true,
+ *   cluster: 'my-cluster',
+ *   onMessage: (event) => console.log('Pod update:', event),
+ *   onError: (error) => console.error('WebSocket error:', error),
+ * });
+ */
+export function useWebSocket<T extends KubeObjectInterface>({
+  url: createUrl,
+  enabled = true,
+  cluster = '',
+  onMessage,
+  onError,
+}: {
+  /** Function that returns the WebSocket URL to connect to */
+  url: () => string;
+  /** Whether the WebSocket connection should be active */
+  enabled?: boolean;
+  /** The Kubernetes cluster ID to watch */
+  cluster?: string;
+  /** Callback function to handle incoming Kubernetes events */
+  onMessage: (data: KubeListUpdateEvent<T>) => void;
+  /** Callback function to handle connection errors */
+  onError?: (error: Error) => void;
+}) {
+  const url = useMemo(() => (enabled ? createUrl() : ''), [enabled, createUrl]);
+  const stableOnMessage = useCallback(
+    (update: any) => {
+      try {
+        if (isKubeListUpdateEvent<T>(update)) {
```

Comment:
useWebsocket hook in theory should handle all kinds of websocket connections, it shouldn't be tightly coupled to list updates. so it shouldn't only pass `KubeListUpdateEvent<T>` but generic data

Reply:
Updated